### ISSUE TITLE
Python3 is preferred. So change #!/usr/bin/env python to #!/usr/bin/env python3?

### ISSUE THREAD

**Issue by strafer** [mdhiggins/sickbeard_mp4_automator#1189]
The wiki states "Python 3 preferred, though Python 2 should be compatible. Future updates are not being actively tested against Python 2."

However, the .py scripts still use /usr/bin/env python in the shebang. As it currently is, I have to do a search and replace of all the scripts to correct the shebang because #!/usr/bin/env python points to Python 2.x.

At least according to this StackOverflow post (https://askubuntu.com/questions/817255/default-to-python3-for-usr-bin-env-python), the correct shebang for Python 3 is #!/usr/bin/env python3.

**Comment by strafer** [comment]
Okay, I'm having way more problems than expected using the new sickbed_mp4_automator and Python 3 on macOS Mojave. It's just not working. Multiple errors. The autoprocess.ini and the logging.ini are not being auto generated when I first run manual.py. 

I think I'm way out of my element here. Maybe I should stick with the last working script and Python 2.7.

**Comment by mdhiggins** [comment]
At work but they are in the /config now

**Comment by mdhiggins** [comment]
Post whatever errors you’re having

**Comment by strafer** [comment]
> At work but they are in the /config now

Do you mean /sickbeard_mp4_automator-master/config? Because if so, that config folder neither exists in the GitHub download nor is created when I run the script.

I tried creating it the config folder inside the automator-master folder but it seemed to make no difference. I'll do some more testing and then post the error messages.

**Comment by cryo-kronos** [comment]
Doing a simple" manual.py -i" command on a sample.mkv generates the following errors:

`Traceback (most recent call last):
  File "/Users/me/Scripts/sickbeard_mp4_automator/manual.py", line 23, in <module>
    log = getLogger("MANUAL")
  File "/Users/me/Scripts/sickbeard_mp4_automator/resources/log.py", line 122, in getLogger
    fileConfig(configfile, defaults={'logfilename': logfile})
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/logging/config.py", line 79, in fileConfig
    handlers = _install_handlers(cp, formatters)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/logging/config.py", line 145, in _install_handlers
    h = klass(*args, **kwargs)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/logging/handlers.py", line 148, in __init__
    BaseRotatingHandler.__init__(self, filename, mode, encoding, delay)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/logging/handlers.py", line 55, in __init__
    logging.FileHandler.__init__(self, filename, mode, encoding, delay)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/logging/__init__.py", line 1087, in __init__
    StreamHandler.__init__(self, self._open())
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/logging/__init__.py", line 1116, in _open
    return open(self.baseFilename, self.mode, encoding=self.encoding)
PermissionError: [Errno 13] Permission denied: '/var/log/sma.log'`

**Comment by Geordon** [comment]
I had that happen yesterday. The following witll make that go away:

sudo touch /var/log/sma.log

sudo chmod o+w /var/log/sma.log


What's happening is that the log file doesn't currently exist and the
directory defaults to root permissions.  The first command creates the file
and the second command makes it so that your user account can write to the
log file.


—
Geordon VanTassle

On March 17, 2020 at 4:29:22 PM, strafer (notifications@github.com) wrote:

Doing a simple" manual.py -i" command on a sample.mkv generates the
following errors:

encoding=self.encoding) PermissionError: [Errno 13] Permission denied:
'/var/log/sma.log'

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/mdhiggins/sickbeard_mp4_automator/issues/1189#issuecomment-600310493>,
or unsubscribe

**Comment by strafer** [comment]
Gordon,

Thank you. That fixed the errors with manual.py.

**Comment by cryo-kronos** [comment]
Now I'm seeing errors when trying to run the postSonarr.py script in Sonarr 3.

`20-3-17 14:21:41.7|Error|postSonarr.py|Traceback (most recent call last):
20-3-17 14:21:41.7|Error|postSonarr.py|  File "/Users/me/Scripts/sickbeard_mp4_automator/postSonarr.py", line 4, in <module>
20-3-17 14:21:41.7|Error|postSonarr.py|    import requests
20-3-17 14:21:41.7|Error|postSonarr.py|ImportError: No module named requests
20-3-17 14:21:41.8|Warn|SonarrErrorPipeline|Invalid request Validation failed: 
 -- : Script exited with code: 1
20-3-17 14:40:39.3|Error|postSonarr.py|Traceback (most recent call last):
20-3-17 14:40:39.3|Error|postSonarr.py|  File "/Users/me/Scripts/sickbeard_mp4_automator/postSonarr.py", line 4, in <module>
20-3-17 14:40:39.3|Error|postSonarr.py|    import requests
20-3-17 14:40:39.3|Error|postSonarr.py|ImportError: No module named requests
20-3-17 14:40:39.3|Warn|SonarrErrorPipeline|Invalid request Validation failed: 
 -- : Script exited with code: 1
20-3-17 14:43:05.9|Warn|SonarrErrorPipeline|Invalid request Validation failed: 
 -- : Script exited with code: 255`

**Comment by Geordon** [comment]
Glad to help!

**Comment by cryo-kronos** [comment]
Forgot to enable debug log:

`20-3-17 14:52:55.0|Debug|Api|[POST] /api/v3/notification/test: 400.BadRequest (25 ms)
20-3-17 14:53:31.7|Debug|CustomScript|Executing external script: /Users/me/Scripts/sickbeard_mp4_automator/postSonarr.py
20-3-17 14:53:31.7|Debug|postSonarr.py|Starting /Users/me/Scripts/sickbeard_mp4_automator/postSonarr.py 
20-3-17 14:53:31.7|Debug|CustomScript|Executed external script: /Users/me/Scripts/sickbeard_mp4_automator/postSonarr.py - Status: 255
20-3-17 14:53:31.7|Debug|CustomScript|Script Output: 

20-3-17 14:53:31.7|Warn|SonarrErrorPipeline|Invalid request Validation failed: 
 -- : Script exited with code: 255
20-3-17 14:53:31.7|Debug|Api|[POST] /api/v3/notification/test: 400.BadRequest (23 ms)`

**Comment by mdhiggins** [comment]
For sonarr 3 you need to make a shell script or batch file to point to your python executable and the script since they’ve removed the argument option. The docker containers generate and include this but for other setups you gotta make it yourself. Otherwise it will execute inside some internal python environment without your dependencies

**Comment by strafer** [comment]
I was using an older version of your script (from a month or two ago) with Sonarr 3 and Python 2 without issue. It's only when I switched to the latest script with it's various changes and Python 3 that the postSonarr.py script stopped working with Sonarr 3.

**Comment by mdhiggins** [comment]
Well it’s failing when it tries to import requests so either you don’t have requests installed or it’s executing via the wrong python environment. So if you changed the shebang around maybe it’s using a different environment.

**Comment by mdhiggins** [comment]
That being said I think I need to make a logging change and have it make a directory in the /var/log folder to avoid permissions issues

**Comment by strafer** [comment]
I'll see if I can figure out and fix my Python 3 and Pip 3 issues so the script starts working.

Where is the autoprocess.ini file supposed to reside now? The config folder?

**Comment by mdhiggins** [comment]
Yeah inside the script root. It should make it for you when you launch manual.py let me know if that’s not the case

**Comment by strafer** [comment]
At least on my system, the manual.py script was not creating the config folder.

**Comment by strafer** [comment]
Python 2 is, for me, easier and more reliable to use than Python 3. Looking over Python 3, it seems that using virtual environments is the preferred way to ensure that things work correctly, but honestly, I think I'm out of my depth. I need a Reddit-style ELI5 breakdown of virtual environments and how to use them. Like how do I get the scripts in mp4_automator to use the virtual environment?

I suspect I'll stick with Python 2 and Sonarr 3 since they work fine with the older version of mp4_automator. 

Once they break, I might have to use GUI tools like iFlicks 3 and SubsMarine and handle things manually, unless I can figure out how to correctly use Python 3 and virtual environments on macOS.

I love using the various scripts and tools on Github, but it seems like Python 3 complicates things beyond my level of competence compared to Python 2.

**Comment by strafer** [comment]
There's even a relevant XKCD comic for what I'm dealing with. 
![python_environment](https://user-images.githubusercontent.com/4935512/76925859-12ec1180-6898-11ea-805d-1d0e0194ac86.png)

**Comment by mdhiggins** [comment]
Its really not that bad and python 2 is EOL. You're welcome to keep using it and I won't intentionally break it but a number of the libraries this script depends on have stopped supporting it and will likely break it in future updates. Virtual environments aren't that complicated and really aren't needed either. If you can install on a virtual environment you can also make your main python environment work too

6579921c2d3bcb904ca951913e1217f9f0833ba9

That should adjust the default logging path to be in the config folder now since /var/log is giving permissions issues
It will also ensure the config folder gets created if you didn't make it

**Comment by mdhiggins** [comment]
26726267e11af9f83a3dd54bfea4bd879d81a1b7
python3 shebang

**Comment by mdhiggins** [comment]
And to answer your question about how to use your virtual environment, you literally just point to your virtual environment executable follow by the python script.
When you open a py file is defaults to python exe defined by your operating system or the shebang, so it runs
/usr/bin/python3 /home/scripts/script.py

To run your virtual environment it would just be
/home/venv/yourrandomenv/python3 /home/scripts/script.py

and when you install your packages in your virtual environment instead of
pip install package

its just 
/home/venv/yourrandomenv/pip install package

**Comment by strafer** [comment]
mdhiggins,

Thanks for the explanation and the updates.

I'm going to nuke my entire homebrew installation (i.e., python, pip, everything) and reinstall homebrew followed by only by python 3, pip 3, and the required dependencies for mp4_automator. Then I'll test it with Sonarr 3 to see if postSonarr.py works with my old config or if I need to create a shell script for it.

I assume for the shell script to work, I just need to have #!/bin/sh one the first line and then the path_to_python followed by the path_to_postSonarr.py, and then make the shell script executable with  chmod +x or chmod 755.

I'll report back my results.

**Comment by kushfest** [comment]
I've also had some kinks since updating to newest version. I'm currently working my way through the needed tweaks to my docker containers that use SMA to to process files: 
sonarr/radarr -> nzbget/deluge -> sonarr/radarr -> plex

It seems that the script is looking for files in a /config folder that doesn't exist or get created. Here's an example of an error I'm seeing when SMA gets run in Deluge:

`06:54:15 [WARNING ][deluge_execute.core           :123 ] stderr: b'Traceback (most recent call last):\n  File "/sma/delugePostProcess.py", line 12, in <module>\n    log = getLogger("DelugePostProcess")\n  File "/sma/resources/log.py", line 114, in getLogger\n    os.mkdir(logpath)\nPermissionError: [Errno 13] Permission denied: \'/sma/config\'\n'`

I'm confused by both the /config folder and then the permission error. Any suggestions?

Meanwhile, in NZBGet, this is the current error that has me stumped:

`ERROR | Tue Mar 17 2020 23:46:27 | Post-process-script  NZBGetPostProcess.py for  Miracle.Workers.2019.S02E08.720p.WEBRip.x264-XLF failed (terminated with  unknown status)
info | Tue Mar 17 2020 23:46:27 | NZBGetPostProcess: KeyError: 'NZBPO_MP4_FOLDER'
info | Tue Mar 17 2020 23:46:27 | NZBGetPostProcess:     raise KeyError(key) from None
info | Tue Mar 17 2020 23:46:27 | NZBGetPostProcess:   File "/usr/lib/python3.6/os.py", line 669, in __getitem__
info | Tue Mar 17 2020 23:46:27 | NZBGetPostProcess:     MP4folder = os.environ['NZBPO_MP4_FOLDER'].strip()
info | Tue Mar 17 2020 23:46:27 | NZBGetPostProcess:   File "/sma/NZBGetPostProcess.py", line 50, in <module>
info | Tue Mar 17 2020 23:46:27 | NZBGetPostProcess: Traceback (most recent call last):
info | Tue Mar 17 2020 23:46:27 | Executing post-process-script NZBGetPostProcess.py for Miracle.Workers.2019.S02E08.720p.WEBRip.x264-XLF`

Any help is greatly appreciated. This script is amazing! Can't wait to get it working again.

**Comment by mdhiggins** [comment]
@kushfest Looks like the first error is because the script doesn't have permission to write to the script directory, you need to add write permission to whatever user you're calling the script from

As a workaround I put a blank file in the config folder so that git includes an empty one for people who don't make the directory. The script will try to create the directory if its absent but without permissions to do so that will fail

e1c5ed7865f58c0296e6438b64163aad476baf5d


For your second error, it looks like

**Comment by mdhiggins** [comment]
https://gist.github.com/mdhiggins/a5ac269dc6d423ea6d237313e8a1347a

There's a generic setup file that will setup everything in macOS when executed. Uncomment the deluge or qbittorrent lines for those dependencies

**Comment by mdhiggins** [comment]
https://gist.github.com/mdhiggins/93f38778b88909fda48f35c19e6b036f

And there's a generic shell script to run sonarr using your venv

**Comment by kushfest** [comment]
Thanks for the response! I'm working now to navigate those changes. I think I'm running up against my own lack of knowledge, but I generally understand. 

The macOS script looked helpful, but didn't quite work when I tried to translate it to my linux-based docker setup. Basically, I am using docker containers for sonarr, radarr, nzbget and deluge that can execute a script as part of their startup process and I use that to install what SMA needs. It's tricky for me to decipher what user is running what process in each docker (just my own lack of experience/knowledge).

The WIP script looks like this:

`apt update
apt install -y --no-install-recommends --no-install-suggests git python3 python3-pip python3-setuptools ffmpeg
if [[ ! -d /sma ]]; then
mkdir /sma
chmod -v -R 775 /sma
curl -fsSL "https://www.github.com/mdhiggins/sickbeard_mp4_automator/tarball/master" | tar xzf - -C "/sma" --strip-components=1
python3 -m pip install --user --upgrade pip
python3 -m pip install --user virtualenv
python3 -m virtualenv /sma/venv
/sma/venv/usr/local/bin/pip install -r /sma/setup/requirements.txt
/sma/venv/usr/local/bin/pip install -r /sma/setup/requirements-deluge.txt
fi
`

No luck at present, I'm getting these errors:
`
/var/run/s6/etc/cont-init.d/mp4_automator.sh: line 14: /sma/venv/usr/local/bin/pip: No such file or directory,
/var/run/s6/etc/cont-init.d/mp4_automator.sh: line 13: /sma/venv/usr/local/bin/pip: No such file or directory`

**Comment by mdhiggins** [comment]
I can help here
Issue is that on linux when you make a python venv it doesn't put the binary in ./venv/usr/local/bin but instead in ./venv/bin. usr/local seems to be a macOS specific thing

```
apt update apt install -y --no-install-recommends --no-install-suggests git python3 python3-pip python3-setuptools ffmpeg if [[ ! -d /sma ]]; then mkdir /sma chmod -v -R 775 /sma curl -fsSL "https://www.github.com/mdhiggins/sickbeard_mp4_automator/tarball/master" | tar xzf - -C "/sma" --strip-components=1 python3 -m pip install --user --upgrade pip python3 -m pip install --user virtualenv python3 -m virtualenv /sma/venv /sma/venv/bin/pip install -r /sma/setup/requirements.txt /sma/venv/bin/pip install -r /sma/setup/requirements-deluge.txt fi
```

Try that

And not sure if you know but I maintain docker containers for sonarr/radarr that work with nzbget/deluge via completed download handling. They're linked in the readme

**Comment by mdhiggins** [comment]
https://gist.github.com/mdhiggins/1d08528b95456e862754fea5431384a7

That's my personal media server docker-compose config if you wanted guidance

**Comment by kushfest** [comment]
Amazing! That solved those errors for me, but the struggle continues. I think it's time for me to migrate to your sonarr/radarr dockers. Your docker compose will be a big help for that. 

My only reservation in the past was that I also needed deluge and nzbget to have sma in their containers to be able to run their respective SMA post-processing scripts. Would that work differently now?

**Comment by mdhiggins** [comment]
So the deluge/nzbget scripts aren't required and the postRadarr/postSonarr scripts can handle everything these days if you enable completed download handling. The file will just be converted after sonarr/radarr rename the file instead but it simplifies the whole setup and means I don't have to maintain as many docker containers

**Comment by kushfest** [comment]
Holy Moly. I love this plan. I'm going to try to start fresh with your sonarr/radarr containers and the standard linuxserver containers for deluge/nzbget. I definitely see how my current setup is too complicated and prone to break when changes come up. 

If I'm reading your docker-compose correctly, it looks like you maintain a folder (/opt/appdata/nzbget/config) with your config files, to be able to use the same config across both sonarr and radarr?

**Comment by mdhiggins** [comment]
Each container needs its own config folder, but the sma portion can share a config folder which you'll see for sonarr/radarr containers is specified

**Comment by kushfest** [comment]
Oh totally. I meant sharing the SMA config only between the dockers. Thanks for confirming.

**Comment by mdhiggins** [comment]
Yeah you can totally do that, you can see my sonarr/radarr containers have the same mounted sma folder

**Comment by kushfest** [comment]
Well I seem to be back in business after moving to your radarr and sonarr containers. I'm just working through some migration issues that aren't SMA related. Thank you again.

**Comment by mdhiggins** [comment]
Gonna close this since the original issues have been resolved but feel free to keep the conversation going @strafer if you need more assistance

**Comment by kushfest** [comment]
Follow-up...this seems to be what's happening:

The downloaded MKV file is getting moved to the final destination before it then gets converted to an MP4. So Radarr is picking up the MKV before the final MP4 is completed by SMA. When SMA does complete, the mkv is removed and the mp4 left behind (I also have a QTFS file, not sure why). Radarr doesn't seem to figure out the change from MKV to MP4 on its own.

What I'm trying to make happen is that the MKV stays where it's downloaded and only the convererted MP4 gets moved to the drive where my movies live. Any idea where I'm going wrong?

**Comment by mdhiggins** [comment]
So unfortunately that’s the downside of not using delugePostProcess etc is that Radarr converts it after it’s done everything else. If you want conversion before then you’ll need to build docker containers that have SMA built in though you could model it after the Radarr and sonarr containers I’ve made. 

It should leave the original file in place and not interrupt seeding though. 

It should also refresh Radarr so that it knows the mp4 is the latest file which you can easily verify. Is there a reason this order doesn’t work?

**Comment by kushfest** [comment]
Oh Ok! It's actually a relief to hear that this is expected behavior. I wasn't seeing any errors, so I was real stumped about where things were going wrong hah. This just seemed less efficient, but I can get used to a new process. In the end, I'd rather stick with the most resilient option and your containers seem to be my best move. Thanks again!

**Comment by mdhiggins** [comment]
It’s really not any less efficient and completed download handling actually tends to have less errors but all the steps that you’re seeing happen regardless
